File: apps/scala/src/main/scala/machinelearning/chapter09/PolynomialFeatures.scala
package machinelearning.chapter09

import machinelearning.chapter02.Features

/** 2 次の項。left と right が同じなら 2 乗の項を表す。
  *
  * @param left
  *   左の列
  * @param right
  *   右の列
  */
case class Term(left: String, right: String):

  /** 項の名前。scikit-learn の get_feature_names_out と同じ形（"RM^2"、"RM LSTAT"）にする。 */
  def name: String = if left == right then s"$left^2" else s"$left $right"

/** 2 次の多項式特徴量（2 乗の項と交互作用の項）を作る。 */
object PolynomialFeatures:

  /** 重複を許して 2 つの列を選ぶ組を、scikit-learn の PolynomialFeatures と同じ順に並べる。 */
  def pairsWithReplacement(columns: Vector[String]): Vector[Term] =
    columns.indices.toVector.flatMap(i => columns.drop(i).map(right => Term(columns(i), right)))

  /** 指定した列の後ろに、2 乗の項と交互作用の項を加える。 */
  def expand(x: Vector[Features], columns: Vector[String]): Vector[Features] =
    val terms = pairsWithReplacement(columns)
    val names = columns ++ terms.map(_.name)
    x.map(features =>
      Features(
        names,
        columns.map(features.value) ++ terms.map(term =>
          features.value(term.left) * features.value(term.right)
        )
      )
    )

  /** 指定した列だけを、その順に選ぶ。 */
  def select(x: Vector[Features], columns: Vector[String]): Vector[Features] =
    x.map(features => Features(columns, columns.map(features.value)))

File: apps/csharp/src/MachineLearning/Chapter09/PolynomialFeatures.cs
namespace MachineLearning.Chapter09;

using MachineLearning.Chapter02;

/// <summary>2 つの列の組。<paramref name="Left"/> と <paramref name="Right"/> が同じなら 2 乗の項を表す。</summary>
/// <param name="Left">左の列</param>
/// <param name="Right">右の列</param>
public sealed record Term(string Left, string Right)
{
    /// <summary>項の名前。scikit-learn の get_feature_names_out と同じ形（"RM^2"・"RM LSTAT"）にする。</summary>
    public string Name => string.Equals(this.Left, this.Right, StringComparison.Ordinal)
        ? $"{this.Left}^2"
        : $"{this.Left} {this.Right}";
}

/// <summary>2 次の多項式特徴量（2 乗の項と交互作用の項）を作る。</summary>
public static class PolynomialFeatures
{
    /// <summary>同じ列どうしの組も含めて、列の組を重複なく作る（[a, b] なら (a,a)・(a,b)・(b,b)）。</summary>
    public static IReadOnlyList<Term> PairsWithReplacement(IReadOnlyList<string> columns)
    {
        ArgumentNullException.ThrowIfNull(columns);
        return [.. columns.SelectMany((left, i) => columns.Skip(i).Select(right => new Term(left, right)))];
    }

    /// <summary>指定した列の後ろに、2 乗の項と交互作用の項を加える。指定しなかった列は残さない。</summary>
    public static IReadOnlyList<Features> Expand(IReadOnlyList<Features> x, IReadOnlyList<string> columns)
    {
        ArgumentNullException.ThrowIfNull(x);
        ArgumentNullException.ThrowIfNull(columns);
        var terms = PairsWithReplacement(columns);
        IReadOnlyList<string> names = [.. columns, .. terms.Select(term => term.Name)];
        return [.. x.Select(features => new Features(
            names,
            [
                .. columns.Select(features.Value),
                .. terms.Select(term => features.Value(term.Left) * features.Value(term.Right)),
            ]))];
    }

    /// <summary>指定した列だけを、その順に選ぶ。</summary>
    public static IReadOnlyList<Features> Select(IReadOnlyList<Features> x, IReadOnlyList<string> columns)
    {
        ArgumentNullException.ThrowIfNull(x);
        ArgumentNullException.ThrowIfNull(columns);
        return [.. x.Select(features => new Features(columns, [.. columns.Select(features.Value)]))];
    }
}
